### DISCUSSION TITLE
Add ArtCNN Shaders

### DISCUSSION THREAD

**Discussion by Zabooby** [dyphire/mpv-config#78] [Q&A] [answered]
They outperform ravu, FSRCNNX and krigbilateral in several tests [here](https://artoriuz.github.io/blog/mpv_upscaling.html).

**Comment by dyphire** [comment, marked as answer]
I have previously tested the effects of the ArtCNN shader, and it shows no significant difference in upscaling of HD content compared to FSRCNNX. It is only slightly stronger in upscaling of SD content than FSRCNNX (both effect poorly, many artifacts). 
There is no reason to add a similar and not significantly improved shader (both trained based on CNN models), Before this, I had already streamlined the shaders in the configuration for similar reasons: https://github.com/dyphire/mpv-config/commit/bb490595aa60e028f7ef2c69a20d11ee553e4c28, https://github.com/dyphire/mpv-config/commit/7cfec9675a0aef6e19159c1fa750fbfe6b03d493
PS: The comparison here is the normal version of ArtCNN, whose DS variant is catastrophic in upscaling SD content.

Don't blindly trust the results of numerical tests, as the ravu ar shader that performed well in that test also performed poorly in upscaling SD content (some artifacts). So you can see that the upscaling shaders used as benchmarks there are not very recommended in my configuration.

Of course, in terms of upscaling high-definition content, the credibility of that numerical test is still good

**Comment by Zabooby** [comment]
By content are you referring to anime or live action? 

Where in your opinion does nnedi3 fit in, similar to FSRCNNX? Good on hd and bad on sd if so why not remove it too? 
Similarly where does FSR's use-case lie?
Would you agree that Anime4k is worse on hd in terms of destroying original detail, FSRCNNX performs better for me, why do you keep it?

Also what are your thoughts on [Cfl](https://github.com/Artoriuz/glsl-chroma-from-luma-prediction) or the chroma variant of ArtCNN as a chroma replacement for krigbilateral? 

My point in asking these questions is to find the best combination of shaders for anime and live action for both sd and hd content without having a shader folder with 20+ shaders.

**Comment by dyphire** [comment]
> By content are you referring to anime or live action?

both, mainly animation.

> Where in your opinion does nnedi3 fit in, similar to FSRCNNX? Good on hd and bad on sd if so why not remove it too? 

The biggest feature of nnedi3 is that it hardly produces artifacts during upsampling and suppresses the original artifacts in the image, making its versatility irreplaceable. It is surprising that after so many years, there is still no better algorithm to replace it.

The reason why FSRCNNX is retained is because the poor upscaling effect of SD content is almost a common problem among other shaders except for specific shaders, and its effects on HD content is indeed good.

> Similarly where does FSR's use-case lie? Would you agree that Anime4k is worse on hd in terms of destroying original detail, FSRCNNX performs better for me, why do you keep it?

The EASU_rgb variant of FSR often effect similarly to nnedi3 and has lower performance consumption, which is why it is retained.

I also do not recommend using Anime4k for HD anime as it will disrupt the original details. But its effects on SD anime is significantly better than any other shader, with powerful anti aliasing and artifact removal capabilities. You may have noticed that the Anime4k usage I recommended in the configuration does not include its upscale algorithm.

https://github.com/dyphire/mpv-config/blob/014c9dbab084cfe5d6081e4ded6dae389229e42f/mpv.conf#L757-L759

> Also what are your thoughts on [Cfl](https://github.com/Artoriuz/glsl-chroma-from-luma-prediction) or the chroma variant of ArtCNN as a chroma replacement for krigbilateral?

I really don't care much about chroma upscaling. The human eye has low sensitivity to chroma information and it's not worth investing too much effort and performance in it. Krigbilateral is also unnecessary. The method of inferring chroma information through luma information using Cfl shaders may not work well when there is no strong correlation between luma and chroma.

But once again, I don't care about chroma upscaling, so I won't test and compare them. The above is just a subjective judgment.

> My point in asking these questions is to find the best combination of shaders for anime and live action for both sd and hd content without having a shader folder with 20+ shaders.

I won't recommend the best combination for anyone because everyone's preferences are different. Shaders with poor effects on SD content have no impact on people who only watch HD content, and for those who like image sharpening, they will not like shaders like nnedi3. So you can only rely on yourself to choose and use the shader scheme that suits you.

**Reply by Zabooby** [reply]
Thanks for the detailed reply, definitely given me a better understanding of the use cases of the various shaders. 

One last question, in a previous issue (#31) you said the hdeband shader was not worth it but since then you've added it to your config, why?

**Reply by dyphire** [reply]
> One last question, in a previous issue (#31) you said the hdeband shader was not worth it but since then you've added it to your config, why?

I personally do not use it, but I think it is worth adding for others to test different deband implementations, just as the reason for adding the PixelClipper shader is also the same.

**Reply by Zabooby** [reply]
@dyphire Would you be open to creating a branch for your personal config, to showcase what you use and not extra stuff added for users of the general config? 

You don't have to update it regularly, maybe once a month.

**Comment by dyphire** [comment]
I turned it into a discussion. Considering that ArtCNN is still actively developing and FSRCNNX is no longer maintained, perhaps one day ArtCNN will be enough to replace FSRCNNX.
~~In fact, I am indeed considering whether to retain only the enhancement variants of FSRCNNX, as they still have merit in SD content.~~

PS: simplify ravu shaders in new commit: https://github.com/dyphire/mpv-config/commit/2c91112bd12eea8545122d4d896581793e56604

**Comment by wathper** [comment]
###  `KrigBilateral.glsl` 对图像可能存在不必要的修改。

对于 `KrigBilateral.glsl` 我做过不少对比。通常情况下，启用它所带来的画面变化是难以察觉的。

但是，当某处局部图像中存在“红色系的、饱和度高的狭小图像”被“饱和度低的、偏灰色的描边”所包围的情况时，它热衷于提高偏灰色那部分图像的饱和度，使该处局部图像变成均一的鲜亮色彩。

这种修改是否正确，我无从判断。基于我自己的喜好，我禁用了 `KrigBilateral.glsl` 

两组示例：
后者为启用 `KrigBilateral.glsl` 之后的效果

1. 红色系字幕的饱和度被提高：
![image1](https://github.com/dyphire/mpv-config/assets/110274576/cb2f498f-15d7-4c1b-bf2b-2fefa8af4c63)
![image2](https://github.com/dyphire/mpv-config/assets/110274576/f1285346-354d-4396-beb9-289d752b9c4f)


2. 红色千纸鹤、红色风车的饱和度被提高。并且风车的灰黑色描边被破坏。
![image3](https://github.com/dyphire/mpv-config/assets/110274576/bb1f6437-d03a-461d-ad0c-1d56193069bc)
![image4](https://github.com/dyphire/mpv-config/assets/110274576/b6344350-4455-42c2-8b39-d66ea5913cfd)

**Reply by dyphire** [reply]
This comparison has no practical reference significance. Only by comparing it with RGB or yuv444 files with complete chroma channel information can we determine which one is closer to the original image.

It should be pointed out that high saturation colors like red will lose saturation after chroma compression, and the krigbilateral shader attempts to restore the lost saturation during chroma upscaling. You can take a look at this extreme example to understand its outstanding performance in this area: [区分不同色度还原算法](https://github.com/hooke007/MPV_lazy/wiki/0_FAQ#%E5%8C%BA%E5%88%86%E4%B8%8D%E5%90%8C%E8%89%B2%E5%BA%A6%E8%BF%98%E5%8E%9F%E7%AE%97%E6%B3%95), at the same time, it may also lead to bleeding effects.

> 风车的灰黑色描边被破坏

This is the potential side effect of such chroma upscaling shaders, which requires selective use by oneself.

**Reply by wathper** [reply]
[区分不同色度还原算法](https://github.com/hooke007/MPV_lazy/wiki/0_FAQ#%E5%8C%BA%E5%88%86%E4%B8%8D%E5%90%8C%E8%89%B2%E5%BA%A6%E8%BF%98%E5%8E%9F%E7%AE%97%E6%B3%95) 这个例子我看过。但在实际观看环境中，能看到红色系图像的明显变化的同时，其他高饱和度的颜色却没有受到什么影响，所以才感觉疑惑。

**Reply by dyphire** [reply]
Therefore, that is an extreme example, where the krigbilateral shader are not simply blindly increasing saturation, but rather attempting to restore lost chromaticity channel information.

**Reply by wathper** [reply]
我原来认为，在相同饱和度下，不同色系的颜色的色度通道信息会被均衡的丢失，实际上红色的色度通道信息更容易被丢失吗？（实际观看中在其他高饱和度的颜色中没有察觉到明显变化）

**Reply by dyphire** [reply]
You need to recognize a practical issue: unlike the tested images, colors in real images do not have the same saturation.

**Reply by wathper** [reply]
“实际观看中在其他高饱和度的颜色中没有察觉到明显变化”指的是不是在同一个画面中，我在很多视频中测试过`KrigBilateral.glsl`，红色部分的变化最容易被观察到。

不过我今天也解开了疑惑。
红色部分的变化最容易被观察到，只是因为我测试过的那些视频中恰好只有红色的色度通道信息被严重丢失了吧。

多谢讲解了。

**Reply by dyphire** [reply]
This is also easy to understand. In chromatography, red is the color with the highest saturation. So red is most likely to lose saturation, and the same goes for HDR. Similarly, it is most prone to bleeding when restoring the saturation.

**Reply by wathper** [reply]
红色是饱和度最高的颜色，这一点我并不知道。如次一来，确实更好理解了。

**Comment by dyphire** [comment]
After some recent testing, I found a better combination of universal shader configurations, see the commits: b9571c5aca7deda036267b16245b74c5b1e6a435, 3742d70d2c49ba9efa97f2fc677b5ae84598b5ac, 4eaace69765e96d6491ccfaba5c5fb565a95b32b

And for animated scenes the best shader currently available is Ani4K. 91a15184913bfb62f6960ba8a4fa13ebf0e27a5c

**Reply by dyphire** [reply]
Let me explain:

The ravu-zoom-ar-r3 shader works like nnedi3 in most test scenario, and the reason for using it is to replace it is because it is directly amplified to the target resolution, so there is no need to consider that the user may use improper scale algorithm to affect the final effect.

The predecessor of the FSRCNNX_x1_16-0-4-1_distort shader is FSRCNNX_x2_16-0-4-1_enhance, which does not perform well, so it is ignored as a successor.
In actual tests, it was found that the effect of FSRCNNX_x1_16-0-4-1_distort is close to that of Anime4K-Restore algorithm, and will not produce additional jagged flaws like Anime4K-Restore, and the impact on the original style is less than that of Anime4K and AniSD.
More importantly, this is a general-purpose  remove compression artifact shader, not limited to animation scenes. It can be said that it is an upper-level replacement of Anime4K-Restore algorithm.
Despite this, my view on FSRCNNX_x1_16-0-4-1_distort is still more suitable for use on low-resolution scenario, and using on high-resolution scenario will still cause loss of details and damage to the original style.

**Reply by dyphire** [reply]
Meanwhile, despite the ArtCNN shader iterations, my view on it has not changed much after testing: 
> and it shows no significant difference in upscaling of HD content compared to FSRCNNX. It is only slightly stronger in upscaling of SD content than FSRCNNX (both effect poorly, many artifacts).

**Reply by Zabooby** [reply]
With the addition of the ArtCNN shaders is there any reason to keep the Ani4K shaders when ArtCNN outperforms it while not completely destroying details like Ani4k?

**Reply by dyphire** [reply]
> With the addition of the ArtCNN shaders is there any reason to keep the Ani4K shaders when ArtCNN outperforms it while not completely destroying details like Ani4k?

Ani4k significantly outperforms ArtCNN in the effect of removing compression artifacts, especially on SD content ArtCNN will only produce deterioration effects. However, because Ani4k's extremely high performance requirements are not suitable for use on real high-definition content, ArtCNN and FSRCNNX should be used.

Edit: The best use scenarios for Ani4k are SD content and poor WEB/Blu-ray distributions. On these contents, Ani4k's powerful ability to remove compression artifacts and retention of the original style make it undoubtedly the best animation shader. But in real HD content, it may lose some subtle details compared to ArtCNN due to its powerful ability to remove compression artifacts.

**Reply by dyphire** [reply]
In fact the Anime4k shader has been completely replaced by Ani4K and FSRCNNX_x1_16-0-4-1_distort, and Anime4k is only worse compared to them.
At one point I wanted to remove the Anime4k shader altogether, but considering the number of users who use and like it, I kept it.

**Reply by Zabooby** [reply]
You should definitely remove them eventually so that the config is more streamlined and easier to maintain. Perhaps make a pinned discussion explaining how Ani4K and FSRCNNX are just better overall.
